File: admin_panel/promo-eligibility.js
/* ═══════════════════════════════════════════════════════════════
   Promo discount eligibility — admin panel copy.
   Source of truth: functions/src/eligibility.ts (SCHEMA.md `promoCodes.eligibility`)
   ═══════════════════════════════════════════════════════════════

   Mirrors functions/src/eligibility.ts and
   customer_app/lib/models/promo_eligibility.dart. This copy is used two ways:
   the create/edit form reads it back out as a plain object to write to
   Firestore (buildEligibility), and the promo table renders it as a summary
   badge (describeEligibility) — it is never used to *gate* a redemption here,
   that authority stays server-side in redeemPromo (P3-03 / PR-5).

   Edit all three, or none. Deliberately no Firebase import — pure functions
   only, so this is testable the same way order-status.js is
   (tools/test-promo-eligibility.mjs, plain `node --test`, no build step). */

export const INELIGIBLE_MESSAGE = {
  not_new_customer: 'This code is for new customers only.',
  not_existing_customer: 'This code is for returning customers only.',
  customer_not_selected: 'This code is not available on your account.',
  merchant_not_eligible: 'This code does not apply to this merchant.',
  category_not_eligible: 'This code does not apply to this category.',
  area_not_eligible: 'This code is not available for your delivery area.',
  not_first_order: 'This code is for your first order only.',
  order_kind_not_eligible: 'This code does not apply to this kind of request.',
};

function norm(s) { return String(s || '').trim().toLowerCase(); }

/** Same rules as functions/src/eligibility.ts checkEligibility — kept here so
 *  the promo table can preview "would this code apply" without a round trip,
 *  and so a future admin-side redemption tool has one place to call. */
export function checkEligibility(elig, ctx) {
  if (!elig) return { eligible: true };

  if (elig.customerScope === 'new' && ctx.priorOrderCount > 0) return fail('not_new_customer');
  if (elig.customerScope === 'existing' && ctx.priorOrderCount <= 0) return fail('not_existing_customer');
  if (elig.customerScope === 'selected') {
    var ids = elig.customerIds || [];
    if (ids.indexOf(ctx.customerId) === -1) return fail('customer_not_selected');
  }

  if (elig.merchantIds && elig.merchantIds.length > 0) {
    if (!ctx.merchantId || elig.merchantIds.indexOf(ctx.merchantId) === -1) return fail('merchant_not_eligible');
  }

  if (elig.categories && elig.categories.length > 0) {
    if (!ctx.merchantCategory || elig.categories.indexOf(ctx.merchantCategory) === -1) return fail('category_not_eligible');
  }

  if (elig.deliveryAreas && elig.deliveryAreas.length > 0) {
    var area = norm(ctx.deliveryArea);
    var matches = area !== '' && elig.deliveryAreas.some(function (a) { return area.indexOf(norm(a)) > -1; });
    if (!matches) return fail('area_not_eligible');
  }

  if (elig.firstOrderOnly && ctx.priorOrderCount > 0) return fail('not_first_order');

  if (elig.orderKinds && elig.orderKinds.length > 0) {
    if (elig.orderKinds.indexOf(ctx.orderKind) === -1) return fail('order_kind_not_eligible');
  }

  return { eligible: true };
}

function fail(reason) { return { eligible: false, reason: reason, message: INELIGIBLE_MESSAGE[reason] }; }

/** Checklist PR-5 "Delivery fee only" / "Order subtotal". */
export function discountBaseAmount(elig, subtotal, deliveryFee) {
  return (elig && elig.discountBase === 'deliveryFee') ? deliveryFee : subtotal;
}

/** Reads the create/edit form's eligibility controls into the plain object
 *  written to `promoCodes/{code}.eligibility`. Returns `null` — not an empty
 *  object — when every control is at its "no restriction" default, so a code
 *  nobody scoped keeps behaving exactly like a pre-PR-5 code (SCHEMA.md).
 *
 *  `pick(id)` reads a <select multiple>'s selected option values; passed in
 *  rather than touching the DOM directly, so this stays pure and testable. */
export function buildEligibility(form) {
  var elig = {};

  if (form.customerScope) elig.customerScope = form.customerScope;
  if (form.customerScope === 'selected' && form.customerIds && form.customerIds.length) {
    elig.customerIds = form.customerIds;
  }
  if (form.merchantIds && form.merchantIds.length) elig.merchantIds = form.merchantIds;
  if (form.categories && form.categories.length) elig.categories = form.categories;
  if (form.deliveryAreas && form.deliveryAreas.length) elig.deliveryAreas = form.deliveryAreas;
  if (form.firstOrderOnly) elig.firstOrderOnly = true;
  if (form.discountBase === 'deliveryFee') elig.discountBase = 'deliveryFee';
  // All three kinds checked (or none touched) means "any kind" — only write
  // orderKinds when the admin has actually narrowed it.
  if (form.orderKinds && form.orderKinds.length && form.orderKinds.length < 3) {
    elig.orderKinds = form.orderKinds;
  }

  return Object.keys(elig).length ? elig : null;
}

/** Turns a comma-separated free-text field into a trimmed, de-duplicated,
 *  non-empty string array — the shared parser behind the Delivery areas
 *  field, so "St. Thomas, , Portland,St. Thomas" reads as two areas. */
export function parseAreas(text) {
  if (!text) return [];
  var seen = {};
  var out = [];
  String(text).split(',').forEach(function (part) {
    var t = part.trim();
    if (!t) return;
    var key = t.toLowerCase();
    if (seen[key]) return;
    seen[key] = true;
    out.push(t);
  });
  return out;
}

/** One short line for the promo table / side panel — "Restricted: new
 *  customers · 2 merchants · St. Thomas" — or null when the code is open to
 *  everyone. Never throws on a malformed/legacy `eligibility` value. */
export function describeEligibility(elig) {
  if (!elig || typeof elig !== 'object') return null;
  var parts = [];
  if (elig.customerScope === 'new') parts.push('new customers');
  else if (elig.customerScope === 'existing') parts.push('existing customers');
  else if (elig.customerScope === 'selected') {
    var n = (elig.customerIds || []).length;
    parts.push(n + ' selected customer' + (n === 1 ? '' : 's'));
  }
  if (elig.firstOrderOnly) parts.push('first order only');
  if (elig.merchantIds && elig.merchantIds.length) parts.push(elig.merchantIds.length + ' merchant' + (elig.merchantIds.length === 1 ? '' : 's'));
  if (elig.categories && elig.categories.length) parts.push(elig.categories.join('/'));
  if (elig.deliveryAreas && elig.deliveryAreas.length) parts.push(elig.deliveryAreas.join(', '));
  if (elig.orderKinds && elig.orderKinds.length) parts.push(elig.orderKinds.join('/'));
  if (elig.discountBase === 'deliveryFee') parts.push('delivery fee only');
  return parts.length ? 'Restricted: ' + parts.join(' · ') : null;
}

File: customer_app/lib/models/promo_eligibility.dart
/// Promo discount eligibility (checklist PR-5).
///
/// Mirrors `functions/src/eligibility.ts` — the server is authoritative; this
/// is the checkout-time preview so a customer sees "this code doesn't apply
/// here" instantly instead of after a round trip. Edit both, or neither.
///
/// "Give me options for who can get the discount" — twelve conditions in the
/// client's own words: all / new / existing / selected customers, selected
/// merchants, selected categories, selected delivery areas, first order only,
/// delivery fee only, order subtotal, Shop & Deliver requests, specific
/// dates/times. The date/time pair is `PromoCode`'s existing `startsAt` /
/// `expiresAt`; everything else lives here.
library;

enum CustomerScope { new_, existing, selected }

/// What kind of request this discount can be redeemed against. `shopDeliver`
/// is a Shop & Deliver quote, not an order.
enum OrderKind { food, package, shopDeliver }

enum DiscountBase { subtotal, deliveryFee }

enum IneligibleReason {
  notNewCustomer,
  notExistingCustomer,
  customerNotSelected,
  merchantNotEligible,
  categoryNotEligible,
  areaNotEligible,
  notFirstOrder,
  orderKindNotEligible,
}

const Map<IneligibleReason, String> ineligibleMessage = {
  IneligibleReason.notNewCustomer: 'This code is for new customers only.',
  IneligibleReason.notExistingCustomer:
      'This code is for returning customers only.',
  IneligibleReason.customerNotSelected:
      'This code is not available on your account.',
  IneligibleReason.merchantNotEligible:
      'This code does not apply to this merchant.',
  IneligibleReason.categoryNotEligible:
      'This code does not apply to this category.',
  IneligibleReason.areaNotEligible:
      'This code is not available for your delivery area.',
  IneligibleReason.notFirstOrder: 'This code is for your first order only.',
  IneligibleReason.orderKindNotEligible:
      'This code does not apply to this kind of request.',
};

class PromoEligibility {
  final CustomerScope? customerScope;
  final List<String>? customerIds;
  final List<String>? merchantIds;
  final List<String>? categories;
  final List<String>? deliveryAreas;
  final bool firstOrderOnly;
  final DiscountBase discountBase;
  final List<OrderKind>? orderKinds;

  const PromoEligibility({
    this.customerScope,
    this.customerIds,
    this.merchantIds,
    this.categories,
    this.deliveryAreas,
    this.firstOrderOnly = false,
    this.discountBase = DiscountBase.subtotal,
    this.orderKinds,
  });

  /// Reads the `eligibility` map stored on a promo document. Absent or
  /// malformed data reads as "no restriction" on every axis — the same
  /// fail-open-to-existing-behaviour choice the server makes, because a promo
  /// created before PR-5 has no `eligibility` field at all.
  factory PromoEligibility.fromMap(Map<String, dynamic>? map) {
    if (map == null) return const PromoEligibility();
    return PromoEligibility(
      customerScope: _scopeOf(map['customerScope']),
      customerIds: _stringList(map['customerIds']),
      merchantIds: _stringList(map['merchantIds']),
      categories: _stringList(map['categories']),
      deliveryAreas: _stringList(map['deliveryAreas']),
      firstOrderOnly: map['firstOrderOnly'] == true,
      discountBase: map['discountBase'] == 'deliveryFee'
          ? DiscountBase.deliveryFee
          : DiscountBase.subtotal,
      orderKinds: _kindList(map['orderKinds']),
    );
  }

  static CustomerScope? _scopeOf(Object? v) => switch (v) {
        'new' => CustomerScope.new_,
        'existing' => CustomerScope.existing,
        'selected' => CustomerScope.selected,
        _ => null,
      };

  static List<String>? _stringList(Object? v) {
    if (v is! List) return null;
    final out = v.whereType<String>().toList();
    return out.isEmpty ? null : out;
  }

  static List<OrderKind>? _kindList(Object? v) {
    if (v is! List) return null;
    final out = v
        .whereType<String>()
        .map((s) => switch (s) {
              'food' => OrderKind.food,
              'package' => OrderKind.package,
              'shop_deliver' => OrderKind.shopDeliver,
              _ => null,
            })
        .whereType<OrderKind>()
        .toList();
    return out.isEmpty ? null : out;
  }
}

class EligibilityContext {
  final String customerId;
  final int priorOrderCount;
  final String? merchantId;
  final String? merchantCategory;
  final String? deliveryArea;
  final OrderKind orderKind;

  const EligibilityContext({
    required this.customerId,
    required this.priorOrderCount,
    this.merchantId,
    this.merchantCategory,
    this.deliveryArea,
    this.orderKind = OrderKind.food,
  });
}

class EligibilityResult {
  final bool eligible;
  final IneligibleReason? reason;
  final String? message;
  const EligibilityResult.ok()
      : eligible = true,
        reason = null,
        message = null;
  const EligibilityResult.rejected(this.reason)
      : eligible = false,
        message = null;
  String? get displayMessage =>
      eligible ? null : (message ?? ineligibleMessage[reason]);
}

/// All specified conditions must pass (AND). See `checkEligibility` in
/// `functions/src/eligibility.ts` for the reasoning behind each one.
EligibilityResult checkEligibility(
  PromoEligibility elig,
  EligibilityContext ctx,
) {
  if (elig.customerScope == CustomerScope.new_ && ctx.priorOrderCount > 0) {
    return const EligibilityResult.rejected(IneligibleReason.notNewCustomer);
  }
  if (elig.customerScope == CustomerScope.existing &&
      ctx.priorOrderCount <= 0) {
    return const EligibilityResult.rejected(
        IneligibleReason.notExistingCustomer);
  }
  if (elig.customerScope == CustomerScope.selected) {
    final ids = elig.customerIds ?? const [];
    if (!ids.contains(ctx.customerId)) {
      return const EligibilityResult.rejected(
          IneligibleReason.customerNotSelected);
    }
  }

  if (elig.merchantIds != null && elig.merchantIds!.isNotEmpty) {
    if (ctx.merchantId == null || !elig.merchantIds!.contains(ctx.merchantId)) {
      return const EligibilityResult.rejected(
          IneligibleReason.merchantNotEligible);
    }
  }

  if (elig.categories != null && elig.categories!.isNotEmpty) {
    if (ctx.merchantCategory == null ||
        !elig.categories!.contains(ctx.merchantCategory)) {
      return const EligibilityResult.rejected(
          IneligibleReason.categoryNotEligible);
    }
  }

  if (elig.deliveryAreas != null && elig.deliveryAreas!.isNotEmpty) {
    final area = (ctx.deliveryArea ?? '').trim().toLowerCase();
    final matches = area.isNotEmpty &&
        elig.deliveryAreas!.any((a) => area.contains(a.trim().toLowerCase()));
    if (!matches) {
      return const EligibilityResult.rejected(IneligibleReason.areaNotEligible);
    }
  }

  if (elig.firstOrderOnly && ctx.priorOrderCount > 0) {
    return const EligibilityResult.rejected(IneligibleReason.notFirstOrder);
  }

  if (elig.orderKinds != null && elig.orderKinds!.isNotEmpty) {
    if (!elig.orderKinds!.contains(ctx.orderKind)) {
      return const EligibilityResult.rejected(
          IneligibleReason.orderKindNotEligible);
    }
  }

  return const EligibilityResult.ok();
}

/// Checklist PR-5 "Delivery fee only" / "Order subtotal".
int discountBaseAmount(PromoEligibility elig, int subtotal, int deliveryFee) =>
    elig.discountBase == DiscountBase.deliveryFee ? deliveryFee : subtotal;
